requests-for-comments/2025-03-14-dw-testing-production-data.md

```diff
@@ -0,0 +1,89 @@
+
+## Types of Actions to Test
+- Syncs
+    - Incremental
+    - Full table
+- HogQl queries (joins)
+
+> Preface: The main idea behind this strategy is to reverse the direction of our standard syncing procedure in order to
+> determine if we can reconstruct our block storage data (delta → s3) back into the original data store it came from (
+> Postgres, MySql, etc…), and query it once it’s reconstructed as before once it has been put back together.
+
+> Note: The diagram below can be used as an aide.
+
+## Steps
+1. The first step would be to transfer data from a configurable list of sub-buckets under
+   `posthog-s3-datawarehouse-us-east-1` into an ephemeral s3 bucket from a list of deltalake tables from prod → dev
+   environment. Using this approach we can avoid looking at credentials and our Postgres instance. This also means that
+   we have very minimal interactions with production as a whole.
+
+2. Reconstruct stateful instances from the delta tables in our CI flow. This would require different processes for
+   various providers. For example, for MySql we could take an official image, deploy it with ArgoCD, then have a program
+   move the data that has been temporarily copied into our dev s3 bucket into entities we consider external data
+   sources.
+
+3. Run an on demand temporal workflow to test our standard syncing operation as if it’s any other production workflow.
+
+4. The results of these workflows will end up back in s3 via delta as an output. We should be able to run queries at
+   this point.
+
+5. Compare the results of the original copied s3 payloads to the s3 payloads that went through the temporal processor
+   and were synced. This is mostly a parity check to ensure that our syncing process is working as expected.
+
+6. Run test case queries on data that has moved from s3 → external_data_source → s3 either with custom queries or test
+   queries inherited from production.
+
+7. Reverse of #1 to clean up all fragments of delta tables that have been copied into the dev environment and buckets
+   that have. Cleanup fragments in Clickhouse that have been generated by this process as well.
+
+[Data Warehouse Testing Production Flow](./images/dw-testing-production-flo.png)
+
+## Out of Scope
+- Temporal Scheduling Logic - We want to ensure that sync works, not that temporal invokes a schedule at the correct
+  interval.
+- Load testing - For massive queries that exceed thresholds encountered in production. This is mostly meant to test if
+  the schemas and queries for syncs are working, not if we can handle massive transactions.
+
+## Tooling
+- ArgoCd
+    - Source containers (Postgres, MySql, Snowflake)
+    - Source destinations (Postgres, MySql, Snowflake)
+    - Clickhouse (Reuse in dev environment as query layer)
+    - Temporal Runner(Reuse in dev environment as query layer)
+- S3
+- GitHub Actions (Invoke when changing subset of our directories, triggering agent)
+
+## Concerns/Open Questions
```

Comment:
Have you considered whether testing becomes a magnitude easier if we were to change our pipeline setup? One think I've been thinking about is how the whole pipeline is a single process, we could split this into two and have an event-based system where we extract data in one workflow, fire the data into a kafka topic, then another process consumes from the topic to load data into S3.

In this model, we'd be able to add an extra consumer on the kafka topic to test PR changes or something similar. My idea isn't fully fleshed out, but would be good to consider different pipeline models here too